File: python/ToyExamples.py
#################################################################################
# ConstrainedPlanningToolbox
# Copyright (C) 2019 Algorithmics group, Delft University of Technology
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#################################################################################


from model.CMDP import CMDP
from model.CPOMDP import CPOMDP
from model.BeliefPoint import BeliefPoint
from instances.CMDPInstance import CMDPInstance
from instances.CPOMDPInstance import CPOMDPInstance
from util.ToolboxServer import ToolboxServer
from algorithms.mdp.constrainedmdp import ConstrainedMDPFiniteHorizon
from algorithms.pomdp.cgcp import CGCP

num_states = 2
num_actions = 2
num_decisions = 10
initial_state = 0

# create reward function
reward_function = [[0.0 for a in range(num_actions)] for s in range(num_states)]
reward_function[1][1] = 10.0

# create transition function
transition_destinations = [[[] for a in range(num_actions)] for s in range(num_states)]
transition_probabilities = [[[] for a in range(num_actions)] for s in range(num_states)]

transition_destinations[0][0].append(0)
transition_probabilities[0][0].append(1.0)

transition_destinations[0][1].append(0)
transition_destinations[0][1].append(1)
transition_probabilities[0][1].append(0.1)
transition_probabilities[0][1].append(0.9)

# define one cost function
num_cost_functions = 1
cost_function = [[[0.0 for a in range(num_actions)] for s in range(num_states)] for k in range(num_cost_functions)]
cost_function[0][1][1] = 2.0

# define cost limits
limits = [0.5]

# create CMDP
cmdp = CMDP(num_states, num_actions, initial_state, num_decisions)
cmdp.set_reward_function(reward_function)
cmdp.set_transitions(transition_destinations, transition_probabilities)
cmdp.set_cost_functions(cost_function)

# create instance
cmdps = []
cmdps.append(cmdp)
cmdp_instance = CMDPInstance(cmdps, num_decisions)
cmdp_instance.set_cost_limits_budget(limits)

# solve the problem
ToolboxServer.connect()
expected_reward = ConstrainedMDPFiniteHorizon.solve(cmdp_instance)
print("Expected reward:", expected_reward)

# extend to POMDP
num_observations = 2
observation_function = [[[0.0 for o in range(num_observations)] for s in range(num_states)] for a in range(num_actions)]
for a in range(num_actions):
    for s_next in range(num_states):
        o = s_next
        observation_function[a][s_next][o] = 1.0

b0 = BeliefPoint([1.0, 0.0])

cpomdp = CPOMDP(num_states, num_actions, num_observations, b0, num_decisions)
cpomdp.set_reward_function(reward_function)
cpomdp.set_transitions(transition_destinations, transition_probabilities)
cpomdp.set_cost_functions(cost_function)
cpomdp.set_observation_function(observation_function)

cpomdps = []
cpomdps.append(cpomdp)
cpomdp_instance = CPOMDPInstance(cpomdps, num_decisions)
cpomdp_instance.set_cost_limits(limits)


expected_reward = CGCP.solve(cpomdp_instance)
print("Expected reward:", expected_reward)

ToolboxServer.disconnect()

File: src/main/java/executables/ToyExamples.java
/*******************************************************************************
 * ConstrainedPlanningToolbox
 * Copyright (C) 2019 Algorithmics group, Delft University of Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *******************************************************************************/
package executables;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import evaluation.CMDPSimulator;
import evaluation.CPOMDPSimulator;

import solutions.mdp.CMDPSolution;
import solutions.pomdp.CPOMDPSolution;

import algorithms.UnsupportedInstanceException;
import algorithms.mdp.constrainedmdp.ConstrainedMDP;
import algorithms.pomdp.CPOMDPAlgorithm;
import algorithms.pomdp.cgcp.CGCP;

import instances.CMDPInstance;
import instances.CPOMDPInstance;
import lp.LPSolver;
import lp.gurobi.LPSolverGurobi;
import model.BeliefPoint;
import model.CMDP;
import model.CPOMDP;

public class ToyExamples {
	public static void main(String[] args) {
		
		LPSolver lpSolver = new LPSolverGurobi();
		Random rnd = new Random(222);
		
		int nStates = 2;
		int nActions = 2;
		int nDecisions = 10;
		int initialState = 0;
				
		
		double[][] rewardFunction = new double[nStates][nActions];
		rewardFunction[1][1] = 10.0;
		
		int[][][] transitionDestinations = new int[nStates][nActions][];
		double[][][] transitionProbabilities = new double[nStates][nActions][];
		
		transitionDestinations[0][0] = new int[]{0};
		transitionProbabilities[0][0] = new double[]{1.0};
		
		transitionDestinations[0][1] = new int[]{0, 1};
		transitionProbabilities[0][1] = new double[]{0.1, 0.9};
		
		transitionDestinations[1][0] = new int[]{0};
		transitionProbabilities[1][0] = new double[]{1.0};
		
		transitionDestinations[1][1] = new int[]{1};
		transitionProbabilities[1][1] = new double[]{1.0};
		
		double[][] costFunction = new double[nStates][nActions];
		costFunction[1][1] = 2.0;
		List<double[][]> costFunctions = new ArrayList<double[][]>();
		costFunctions.add(costFunction);
		
		CMDP cmdp = new CMDP(nStates, nActions, costFunctions, initialState, nDecisions);
		cmdp.setRewardFunction(rewardFunction);
		cmdp.setTransitionFunction(transitionDestinations, transitionProbabilities);
		
		CMDPInstance cmdpInstance = CMDPInstance.createBudgetInstance(new CMDP[]{cmdp}, new double[]{0.5}, nDecisions);
		
		
		ConstrainedMDP mdpAlgorithm = new ConstrainedMDP(lpSolver, rnd);
		try {
			mdpAlgorithm.setInstance(cmdpInstance);
		} catch (UnsupportedInstanceException e) {
			e.printStackTrace();
		}
		
		CMDPSolution mdpSolution = mdpAlgorithm.solve();
		
		
		CMDPSimulator mdpSim = new CMDPSimulator(cmdpInstance, mdpSolution, rnd);
		mdpSim.run(2000000);
		System.out.println("Mean reward: "+mdpSim.getMeanReward());
		System.out.println("Mean cost: "+mdpSim.getMeanTotalCost(0));
		
		
		
		
		int nObservations = 2;
		double[][][] observationFunction = new double[nActions][nStates][nObservations];
		
		for(int a=0; a<nActions; a++) {
			for(int sNext=0; sNext<nStates; sNext++) {
				int o = sNext;
				observationFunction[a][sNext][o] = 1.0;
			}
		}
		
		BeliefPoint b0 = new BeliefPoint(new double[]{1.0, 0.0});
		
		CPOMDP cpomdp = new CPOMDP(nStates, nActions, nObservations, costFunctions, observationFunction, b0, nDecisions);
		cpomdp.setRewardFunction(rewardFunction);
		cpomdp.setTransitionFunction(transitionDestinations, transitionProbabilities);
		
		CPOMDPInstance cpomdpInstance = CPOMDPInstance.createBudgetInstance(new CPOMDP[]{cpomdp}, new double[]{0.5}, nDecisions);
		
		CPOMDPAlgorithm pomdpAlgorithm = new CGCP(lpSolver, rnd);
		try {
			pomdpAlgorithm.setInstance(cpomdpInstance);
		} catch (UnsupportedInstanceException e) {
			e.printStackTrace();
			System.exit(0);
		}	
		CPOMDPSolution pomdpSolution = pomdpAlgorithm.solve();
		
		CPOMDPSimulator pomdpSim = new CPOMDPSimulator(cpomdpInstance, pomdpSolution, rnd);
		pomdpSim.run(2000000);
		System.out.println("Mean reward: "+pomdpSim.getMeanReward());
		System.out.println("Mean cost: "+pomdpSim.getMeanTotalCost(0));
	}
}
